db/view/view.cc

```diff
@@ -1240,6 +1240,14 @@ future<> mutate_MV(
                 }
             }
         }
+        // It's still possible that a target endpoint is dupliated in the remote endpoints list,
+        // so let's get rid of the duplicate if it exists
+        if (target_endpoint) {
+            auto remote_it = std::find(remote_endpoints.begin(), remote_endpoints.end(), *target_endpoint);
+            if (remote_it != remote_endpoints.end()) {
+                remote_endpoints.erase(remote_it);
+            }
+        }
```

Comment:
Curious, why would there be duplicates in the first place?

Reply:
I mentioned that it another thread - I'd love to know... I'm now trying to figure out if `get_natural_endpoints` can return something that's also returned from `pending_endpoints_for` - that's the only source of duplicates. I don't really grasp this code, so it's hard to reason about it. It is however the only way I see for a write handler to end up breaking a promise. All the other code paths correctly propagate either success or exception.

Reply:
I think we discussed this a lot in the last iteration of trying to fix this same problem ;-) I don't know exactly why a node can appear on both natural_endpoint and pending_endpoint - it might be just the normal fact that gossip takes time, so nothing is every perfectly synchronized. However, there is no reason to assume that if this problem happens it can only happen for the *local* node (if anything, it's even stranger when it happens to the local node - how can a node not be synchronized on its own state?).

Reply:
I will commit this PR. I share Avi's frustration that we never understand exactly when this duplication happens. But clearly, it does happen, and when it does, removing it is indeed the correct thing - it can't hurt not to send the same message to the same node twice.

Reply:
The root of the problem is that it actually gets sent once - because this vector of endpoints becomes an unordered set during the processing. Unfortunately, the response handlers are created based on the vector, so we can create two handlers for a single message, and one of them never sees its response and dies, alone and forgotten, producing a broken promise.

Reply:
@asias any idea why a node can appear on both pending_endpoint and natural_endpoint?

Reply:
This will happen when we do replacing operation with replacing node takes write enabled

We switch to use get_natural_endpoints_without_node_being_replaced for normal read/write.

But we can not use that in MV, because it will generate wrong base and view table paring.

I had a local patch recently. 

```
--- a/db/view/view.cc
+++ b/db/view/view.cc
@@ -1161,7 +1161,13 @@ get_view_natural_endpoint(const sstring& keyspace_name,
         // This node is not a base replica of this key, so we return empty
         return {};
     }
-    return view_endpoints[base_it - base_endpoints.begin()];
+    auto node = view_endpoints[base_it - base_endpoints.begin()];
+    // If the selected view node is being replaced, skip it here because it
+    // will be selected from pending_endpoints.
+    if (rs.is_being_replaced(node)) {
+        return {};
+    }
+    return node;
 }
```

Reply:
btw @nyh @psarna I sent you email regarding possible mv write duplication a few days ago. Can you please take a look?